File: crates/tauri-plugin-gattify/android/src/main/java/dev/gattify/plugin/ExecuteResult.kt
package dev.gattify.plugin

import app.tauri.plugin.JSObject

internal sealed interface ExecuteResult {
  class Resolve(val reply: JSObject) : ExecuteResult
  class Reject(val message: String, val code: String) : ExecuteResult
}

/** The adapter facts that the status commands read. */
internal interface StatusProbe {
  val hasAdapter: Boolean
  val adapterOn: Boolean

  /** `BLUETOOTH_CONNECT` is granted on API 31 and later. Always true before. */
  val connectPermitted: Boolean
  val hasAdvertiser: Boolean

  fun permissions(): PermissionOutcomes
}

/**
 * Answers a status command at once and rejects an unknown command. Returns null
 * for every other command: the backend runs it.
 */
internal fun executeResult(kind: String?, probe: StatusProbe): ExecuteResult? {
  if (kind == null || kind !in COMMAND_KINDS) {
    return ExecuteResult.Reject(
      "the Android backend does not know ${if (kind.isNullOrEmpty()) "this command" else kind}",
      ErrorCode.UNSUPPORTED,
    )
  }
  return when (kind) {
    "getState" -> ExecuteResult.Resolve(Replies.state(adapterState(probe)))
    "getCapabilities" -> ExecuteResult.Resolve(
      Replies.capabilities(CapabilityFacts(probe.hasAdapter, probe.adapterOn, probe.hasAdvertiser)),
    )
    "checkPermissions" -> ExecuteResult.Resolve(Replies.permissions(probe.permissions()))
    else -> null
  }
}

internal fun adapterState(probe: StatusProbe): String = when {
  !probe.hasAdapter -> "unavailable"
  !probe.connectPermitted -> "unauthorized"
  probe.adapterOn -> "poweredOn"
  else -> "poweredOff"
}

File: crates/tauri-plugin-gattify/ios/Sources/ExecuteResult.swift
import Tauri

enum ExecuteResult {
  case resolve(JsonObject)
  case reject(message: String, code: String)
}

func executeResult(kind: String?, adapterState: () -> String) -> ExecuteResult {
  switch kind {
  case "getState":
    return .resolve(reply("state", adapterState()))
  case "getCapabilities":
    return .resolve(reply("capabilities", capabilities()))
  case "checkPermissions":
    return .resolve(
      reply("permissions", ["scan": "unknown", "connect": "unknown", "advertise": "unknown"]))
  case "cancel", "closeOwner":
    return .resolve(reply("empty"))
  default:
    let name = kind.flatMap { $0.isEmpty ? nil : $0 } ?? "this command"
    return .reject(message: "the iOS backend does not implement \(name) yet", code: "unsupported")
  }
}

private func reply(_ kind: String, _ payload: Any? = nil) -> JsonObject {
  var reply: JsonObject = ["kind": kind]
  if let payload {
    reply["payload"] = payload
  }
  return reply
}

private func capabilities() -> JsonObject {
  let notImplemented: JsonObject = ["level": "unknown", "reason": "backendNotImplemented"]
  return [
    "central": notImplemented,
    "peripheral": notImplemented,
    "advertising": notImplemented,
    "targetedNotify": notImplemented,
    "simultaneousRoles": notImplemented,
    "background": ["level": "unsupported", "reason": "foregroundOnlyContract"] as JsonObject,
  ]
}
